src/aspire/nfft/cufinufft.py

```diff
@@ -0,0 +1,131 @@
+        :param many: Optional integer indicating if you would like to compute a batch of `many`
+        transforms.  Implies vol_f.shape is (..., `many`). Defaults to 0 which disables batching.
+        """
+
+        # Passing "many" expects one large higher dimensional array.
+        #   Set some housekeeping variables so we can discern how to handle the dims later.
+        self.many = False
+        self.ntransforms = 1
+        if many != 0:
+            self.many = True
+            self.ntransforms = many
+
+        # Basic dtype passthough.
+        dtype = fourier_pts.dtype
+        if dtype == np.float64 or dtype == np.complex128:
+            self.dtype = np.float64
+            self.complex_dtype = np.complex128
+        elif dtype == np.float32 or dtype == np.complex64:
+            self.dtype = np.float32
+            self.complex_dtype = np.complex64
+        else:
+            raise RuntimeError("Unsupported dtype encountered")
+
+        self.sz = sz
+        self.dim = len(sz)
+        # TODO: Following the row major conversion, this should no longer happen.
+        if fourier_pts.flags.f_contiguous:
+            logger.warning('cufinufft has caught an F_CONTIGUOUS array,'
+                           ' `fourier_pts` will be copied to C_CONTIGUOUS.')
+        self.fourier_pts = np.asarray(np.mod(fourier_pts + np.pi, 2 * np.pi) - np.pi, order='C', dtype=self.dtype)
+        self.num_pts = fourier_pts.shape[1]
+        self.epsilon = max(epsilon, np.finfo(self.dtype).eps)
+
+        self._transform_plan = cufinufft(2, self.sz, -1, self.epsilon, ntransforms=self.ntransforms,
+                                         dtype=self.dtype)
+
+        self._adjoint_plan = cufinufft(1, self.sz, 1, self.epsilon, ntransforms=self.ntransforms,
+                                       dtype=self.dtype)
+
+    def transform(self, signal):
+        """
+        Compute the NUFFT transform using this plan instance.
+
+        :param signal: Signal to be transformed. For a single transform,
+        this should be a a 1, 2, or 3D array matching the plan `sz`.
+        For a batch, signal should have shape `(*sz, many)`.
+
+        :returns: Transformed signal of shape `num_pts` or
+        `(many, num_pts)`.
+        """
+
+        assert signal.dtype == self.dtype or signal.dtype == self.complex_dtype
+
+        sig_shape = signal.shape
+        if self.many:
+            sig_shape = signal.shape[:-1]         # order...
+        ensure(sig_shape == self.sz, f'Signal to be transformed must have shape {self.sz}')
+
+        # Note that it would be nice to address this ordering enforcement after the C major conversions.
+        signal_gpu = gpuarray.to_gpu(signal.astype(self.complex_dtype, copy=False, order='F'))
+
+        # This ordering situation is a little strange, but it works.
+        result_gpu = gpuarray.GPUArray((self.ntransforms, self.num_pts), dtype=self.complex_dtype, order='C')
```

Comment:
Can we add the notation why the signal_gpu and result_gpu need to have different orders? Is this because of inconsistency inside C++ code?

Reply:
No, its because ASPIRE code is a hot mess. This enforces what comes in and out will work in ASPIRE until ASPIRE is improved.

I added some details around, and some traps for problems.  Largely I didn't add lots of details because it would pick certain peoples work, and its not as good a use of my time relative to just fixing things.  This code will need a second pass after the ordering changes (but it should be relatively minor, not like this).